pkg/session/session.go

```diff
@@ -1146,6 +1146,7 @@ func (s *session) retry(ctx context.Context, maxCnt uint) (err error) {
 		for i, sr := range nh.history {
 			st := sr.st
 			s.sessionVars.StmtCtx = sr.stmtCtx
+			s.sessionVars.StmtCtx.CTEStorageMap = map[int]*executor.CTEStorages{}
```

Comment:
I believe this missing code is a typical example of 'session context' issue of TiDB. /cc @lcwangchao

Reply:
But I think the actual reason is the code path in retry is different with executing a new statement normally. In normal cases, `ResetContextOfStmt` is called when we enter a new statement, but in retry logic, we use `ResetForRetry` instead.

Reply:
Yes, but we cannot totally reset the stmtctx for retry.

Reply:
The retry behavior is disabled by default.
It cause lost update https://github.com/pingcap/tidb/issues/10075

Reply:
I really hope one day we can get rid of the legacy code like this.
We can simplify the logic a lot!